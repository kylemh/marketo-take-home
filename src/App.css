.App h1 {
  text-align: center;
}

.textAreas {
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
}

.textAreas > p {
  margin: 2rem;
}

.textAreaContainer {
  margin: 1rem;
}

.textAreaContainer > label {
  display: block;
  font-size: 1.25rem;
  font-weight: bold;
  margin-top: 1rem;
  margin-bottom: 1rem;
}

.errorBox {
  color: red;
  max-width: 320px;
  height: 20px;
}

.buttonContainer {
  flex: 1 1 100%;
  text-align: center;
}

.button {
  background-color: rgba(89, 68, 176, 1);
  border: 1px solid rgba(0, 0, 0, 0);
  color: #ffffff;
  font-size: 1rem;
  border-radius: 4px;
  margin: 1rem auto;
  padding: 1rem 1.25rem;
}

.button:hover {
  background-color: rgba(89, 68, 176, 0.9);
  border: 1px solid rgba(0, 0, 0, 0.5);
  cursor: pointer;
}

.logArea {
  margin: 1rem auto;
  max-width: 675px;
}

@media screen and (max-width: 675px) {
  .logArea {
    margin: 1rem 2rem;
  }
}
