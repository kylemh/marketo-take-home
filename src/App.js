import React, { Component } from 'react';
import dedupeLeads from './utils/dedupeLeads';
import './App.css';

class App extends Component {
  state = {
    errorMessage: '',
    inputValue: '',
    outputValue: '',
    removedLeads: [],
  };

  dedupeInput = (ev) => {
    const { state } = this;
    ev.preventDefault();

    try {
      const leads = JSON.parse(state.inputValue)['leads'];

      if (leads) {
        const { dedupedLeads, removedLeads } = dedupeLeads(leads);

        this.setState({
          outputValue: JSON.stringify({ leads: dedupedLeads }, undefined, 2),
          errorMessage: '',
          removedLeads,
        });
      }
    } catch {
      this.setState({
        errorMessage:
          'The input must be valid JSON and include a key "leads" with an array as a value.',
      });
    }
  };

  formatJSON = (ev) => {
    const value = ev.target.value;

    try {
      const parsedJSON = JSON.parse(value);

      if (parsedJSON) {
        this.setState({
          inputValue: JSON.stringify(parsedJSON, undefined, 2),
          errorMessage: '',
        });
      }
    } catch (error) {
      this.setState({ inputValue: ev.target.value, errorMessage: error.message });
    }
  };

  updateInputValue = (ev) => {
    const inputValue = ev.target.value;

    this.setState({ inputValue });
  };

  render() {
    const { state } = this;

    return (
      <main className="App">
        <h1>Marketo Take-Home</h1>

        <form className="textAreas" onSubmit={this.dedupeInput}>
          <p>Put JSON of leads into input, press dedupe, and watch the magic!</p>

          <div className="buttonContainer">
            <button className="button" type="submit">
              Dedupe!
            </button>
          </div>

          <section className="textAreaContainer">
            <label>Input:</label>
            <textarea
              cols="50"
              rows="20"
              onBlur={this.formatJSON}
              onChange={this.updateInputValue}
              value={state.inputValue}
              name="input"
            />
            <p className="errorBox">{state.errorMessage}</p>
          </section>

          <section className="textAreaContainer">
            <label>Output:</label>
            <textarea cols="50" rows="20" readOnly value={state.outputValue} name="output" />
          </section>
        </form>

        <section className="logArea">
          {state.removedLeads.length > 0 && <p>Removed:</p>}

          {/* Key as index here isn't necessary, but it's in here to prevent a lint warning 🤷‍♂️ */}
          {state.removedLeads.map((removedLead, index) => (
            <pre key={index}>{JSON.stringify(removedLead, undefined, 2)}</pre>
          ))}
        </section>
      </main>
    );
  }
}

export default App;
