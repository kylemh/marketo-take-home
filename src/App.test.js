import React from 'react';
import ReactDOM from 'react-dom';
import { mount } from 'enzyme';
import App from './App';

function dedupeLeadsInUI(wrapper, leads) {
  const inputTextarea = wrapper.find('textarea[name="input"]');
  const json = JSON.stringify({ leads });

  inputTextarea.simulate('change', { target: { value: json } });

  wrapper.update();

  const button = wrapper.find('button');
  button.simulate('submit');
}

describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should change the error message if input is empty and dedupe button is clicked', () => {
    const wrapper = mount(<App />);

    const button = wrapper.find('button');
    button.simulate('submit');

    wrapper.update();

    const errorBox = wrapper.find('.errorBox');
    expect(errorBox.text()).toStrictEqual(
      'The input must be valid JSON and include a key "leads" with an array as a value.',
    );
  });

  it('should change output to copy input if it is one lead', () => {
    const leads = [
      {
        _id: 'jkj238238jdsnfsj23',
        email: 'foo@bar.com',
        firstName: 'John',
        lastName: 'Smith',
        address: '123 Street St',
        entryDate: '2014-05-07T17:30:20+00:00',
      },
    ];

    const wrapper = mount(<App />);
    dedupeLeadsInUI(wrapper, leads);

    const outputTextArea = wrapper.find('textarea[name="output"]');
    expect(outputTextArea.text()).toStrictEqual(JSON.stringify({ leads }, undefined, 2));
  });

  it('should display event log after deduping leads with dupes', () => {
    const _id = 'jkj238238jdsnfsj23';

    const betterLead = {
      _id,
      email: 'yo@ya.com',
      firstName: 'Eldritch',
      lastName: 'Smacklebock',
      address: '123 Street Avenue',
      entryDate: '2014-06-07T17:30:20+00:00',
    };

    const worseLead = {
      _id,
      email: 'foo@bar.com',
      firstName: 'John',
      lastName: 'Smith',
      address: '123 Street St',
      entryDate: '2014-05-07T17:30:20+00:00',
    };

    const leadsWithDupes = [worseLead, betterLead];

    const wrapper = mount(<App />);

    dedupeLeadsInUI(wrapper, leadsWithDupes);

    expect(wrapper.find('.logArea').text()).toContain('Removed:');
    expect(wrapper.find('.logArea').text()).toContain(worseLead.firstName);
  });
});
