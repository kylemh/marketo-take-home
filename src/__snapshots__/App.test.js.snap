// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<App>
  <main
    className="App"
  >
    <h1>
      Marketo Take-Home
    </h1>
    <form
      className="textAreas"
      onSubmit={[Function]}
    >
      <p>
        Put JSON of leads into input, press dedupe, and watch the magic!
      </p>
      <div
        className="buttonContainer"
      >
        <button
          className="button"
          type="submit"
        >
          Dedupe!
        </button>
      </div>
      <section
        className="textAreaContainer"
      >
        <label>
          Input:
        </label>
        <textarea
          cols="50"
          name="input"
          onChange={[Function]}
          rows="20"
          value=""
        />
        <p
          className="errorBox"
        />
      </section>
      <section
        className="textAreaContainer"
      >
        <label>
          Output:
        </label>
        <textarea
          cols="50"
          name="output"
          readOnly={true}
          rows="20"
          value=""
        />
      </section>
    </form>
    <section
      className="logArea"
    />
  </main>
</App>
`;
